src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -691,23 +695,23 @@ internal static bool Dispatch()
                     //
                     // Execute the workitem outside of any finally blocks, so that it can be aborted if needed.
                     //
+#if FEATURE_OBJCWRAPPERS
+                    if (ThreadPool.EnableDispatchAutoReleasePool)
+                    {
+#pragma warning disable CA1416 // This call site is reachable on all platforms. 'Wrappers.DrainAutoreleasePool(IntPtr)' is only supported on: 'macos'. This code is only compiled on macos.
+                        DispatchItemWithAutoReleasePool(workItem, currentThread);
+#pragma warning restore CA1416
+                    }
+                    else
+#endif
 #pragma warning disable CS0162 // Unreachable code detected. EnableWorkerTracking may be a constant in some runtimes.
                     if (ThreadPool.EnableWorkerTracking)
                     {
                         DispatchWorkItemWithWorkerTracking(workItem, currentThread);
                     }
-                    else if (workItem is Task task)
-                    {
-                        // Check for Task first as it's currently faster to type check
-                        // for Task and then Unsafe.As for the interface, rather than
-                        // vice versa, in particular when the object implements a bunch
-                        // of interfaces.
-                        task.ExecuteFromThreadPool(currentThread);
-                    }
                     else
                     {
-                        Debug.Assert(workItem is IThreadPoolWorkItem);
-                        Unsafe.As<IThreadPoolWorkItem>(workItem).Execute();
+                        DispatchWorkItem(workItem, currentThread);
```

Comment:
It hopefully shouldn't, but let's make absolute sure this doesn't regress anything on Linux and Windows.

Reply:
I agree. Do we have a good way to run the perf tests locally to validate that we don't reduce perf here?

Reply:
You can build a runtime locally and then submit a request to the asp.net perf lab and your runtime libs. I will need to look up the command tomorrow.

Reply:
This is the page that lists all the commands:
https://github.com/aspnet/Benchmarks/tree/master/scenarios
It'd be good to check a few, like `plaintext`, `json`, and `fortunes`.

Reply:
Here's some results from running those benchmarks against master and this PR . It looks like this PR is faster in `plaintext` and `fortunes` and worse in `json`, but it might just be due to the usual instability in networking-based benchmarks:



# Runs against .NET 6 Master

## plaintext

| application           |                                                                   |
| --------------------- | ----------------------------------------------------------------- |
| CPU Usage (%)         | 88                                                                |
| Cores usage (%)       | 1,051                                                             |
| Working Set (MB)      | 49                                                                |
| Build Time (ms)       | 3,264                                                             |
| Start Time (ms)       | 287                                                               |
| Published Size (KB)   | 103,364                                                           |
| .NET Core SDK Version | 6.0.100-preview.2.21118.23                                        |
| ASP.NET Core Version  | 6.0.0-preview.2.21118.14+7a0e7fef3b4d32a42a014e6de2652034a37952f5 |
| .NET Runtime Version  | 6.0.0-preview.2.21118.10+1db079d19538a7520ec6c74f0566b4b78c7e935d |


| load                   |            |
| ---------------------- | ---------- |
| CPU Usage (%)          | 54         |
| Cores usage (%)        | 646        |
| Working Set (MB)       | 37         |
| Start Time (ms)        | 0          |
| First Request (ms)     | 107        |
| Requests/sec           | 2,363,904  |
| Requests               | 35,694,801 |
| Mean latency (ms)      | 2.08       |
| Max latency (ms)       | 229.60     |
| Bad responses          | 0          |


## json

| application           |                                                                   |
| --------------------- | ----------------------------------------------------------------- |
| CPU Usage (%)         | 78                                                                |
| Cores usage (%)       | 934                                                               |
| Working Set (MB)      | 152                                                               |
| Build Time (ms)       | 3,436                                                             |
| Start Time (ms)       | 296                                                               |
| Published Size (KB)   | 103,364                                                           |
| .NET Core SDK Version | 6.0.100-preview.2.21118.23                                        |
| ASP.NET Core Version  | 6.0.0-preview.2.21118.14+7a0e7fef3b4d32a42a014e6de2652034a37952f5 |
| .NET Runtime Version  | 6.0.0-preview.2.21118.10+1db079d19538a7520ec6c74f0566b4b78c7e935d |


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 68        |
| Cores usage (%)        | 816       |
| Working Set (MB)       | 38        |
| Start Time (ms)        | 1         |
| First Request (ms)     | 125       |
| Requests/sec           | 350,482   |
| Requests               | 5,292,163 |
| Mean latency (ms)      | 1.95      |
| Max latency (ms)       | 202.26    |
| Bad responses          | 0         |
| Socket errors          | 0         |
| Read throughput (MB/s) | 50.81     |
| Latency 50th (ms)      | 0.69      |
| Latency 75th (ms)      | 0.81      |
| Latency 90th (ms)      | 0.96      |
| Latency 99th (ms)      | 48.80     |
| Socket errors          | 0          |
| Read throughput (MB/s) | 297.58     |
| Latency 50th (ms)      | 0.97       |
| Latency 75th (ms)      | 1.39       |
| Latency 90th (ms)      | 1.67       |
| Latency 99th (ms)      | 0.00       |

## fortunes

| db               |        |
| ---------------- | ------ |
| CPU Usage (%)    | 57     |
| Cores usage (%)  | 683    |
| Working Set (MB) | 280    |
| Build Time (ms)  | 80,195 |
| Start Time (ms)  | 417    |


| application           |                                                                   |
| --------------------- | ----------------------------------------------------------------- |
| CPU Usage (%)         | 92                                                                |
| Cores usage (%)       | 1,107                                                             |
| Working Set (MB)      | 197                                                               |
| Build Time (ms)       | 5,172                                                             |
| Start Time (ms)       | 359                                                               |
| Published Size (KB)   | 103,364                                                           |
| .NET Core SDK Version | 6.0.100-preview.2.21118.23                                        |
| ASP.NET Core Version  | 6.0.0-preview.2.21118.14+7a0e7fef3b4d32a42a014e6de2652034a37952f5 |
| .NET Runtime Version  | 6.0.0-preview.2.21118.10+1db079d19538a7520ec6c74f0566b4b78c7e935d |


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 30        |
| Cores usage (%)        | 356       |
| Working Set (MB)       | 37        |
| Start Time (ms)        | 1         |
| First Request (ms)     | 552       |
| Requests/sec           | 109,357   |
| Requests               | 1,651,298 |
| Mean latency (ms)      | 2.41      |
| Max latency (ms)       | 162.34    |
| Bad responses          | 0         |
| Socket errors          | 0         |
| Read throughput (MB/s) | 142.57    |
| Latency 50th (ms)      | 2.27      |
| Latency 75th (ms)      | 2.58      |
| Latency 90th (ms)      | 2.93      |
| Latency 99th (ms)      | 6.95      |

# Runs against this branch

## plaintext

| application           |                                                                   |
| --------------------- | ----------------------------------------------------------------- |
| CPU Usage (%)         | 88                                                                |
| Cores usage (%)       | 1,052                                                             |
| Working Set (MB)      | 49                                                                |
| Build Time (ms)       | 3,238                                                             |
| Start Time (ms)       | 326                                                               |
| Published Size (KB)   | 103,364                                                           |
| .NET Core SDK Version | 6.0.100-preview.2.21118.23                                        |
| ASP.NET Core Version  | 6.0.0-preview.2.21118.14+7a0e7fef3b4d32a42a014e6de2652034a37952f5 |
| .NET Runtime Version  | 6.0.0-dev                                                         |


| load                   |            |
| ---------------------- | ---------- |
| CPU Usage (%)          | 56         |
| Cores usage (%)        | 670        |
| Working Set (MB)       | 38         |
| Start Time (ms)        | 1          |
| First Request (ms)     | 108        |
| Requests/sec           | 2,452,144  |
| Requests               | 37,024,758 |
| Mean latency (ms)      | 2.03       |
| Max latency (ms)       | 235.87     |
| Bad responses          | 0          |
| Socket errors          | 0          |
| Read throughput (MB/s) | 308.69     |
| Latency 50th (ms)      | 0.95       |
| Latency 75th (ms)      | 1.36       |
| Latency 90th (ms)      | 1.63       |
| Latency 99th (ms)      | 0.00       |

## json

| application           |                                                                   |
| --------------------- | ----------------------------------------------------------------- |
| CPU Usage (%)         | 76                                                                |
| Cores usage (%)       | 915                                                               |
| Working Set (MB)      | 152                                                               |
| Build Time (ms)       | 5,143                                                             |
| Start Time (ms)       | 339                                                               |
| Published Size (KB)   | 103,364                                                           |
| .NET Core SDK Version | 6.0.100-preview.2.21118.23                                        |
| ASP.NET Core Version  | 6.0.0-preview.2.21118.14+7a0e7fef3b4d32a42a014e6de2652034a37952f5 |
| .NET Runtime Version  | 6.0.0-dev                                                         |


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 67        |
| Cores usage (%)        | 806       |
| Working Set (MB)       | 37        |
| Start Time (ms)        | 0         |
| First Request (ms)     | 125       |
| Requests/sec           | 344,970   |
| Requests               | 5,208,814 |
| Mean latency (ms)      | 1.96      |
| Max latency (ms)       | 311.77    |
| Bad responses          | 0         |
| Socket errors          | 0         |
| Read throughput (MB/s) | 50.01     |
| Latency 50th (ms)      | 0.70      |
| Latency 75th (ms)      | 0.82      |
| Latency 90th (ms)      | 0.98      |
| Latency 99th (ms)      | 47.55     |

## fortunes

| db               |       |
| ---------------- | ----- |
| CPU Usage (%)    | 60    |
| Cores usage (%)  | 717   |
| Working Set (MB) | 311   |
| Build Time (ms)  | 2,093 |
| Start Time (ms)  | 389   |


| application           |                                                                   |
| --------------------- | ----------------------------------------------------------------- |
| CPU Usage (%)         | 92                                                                |
| Cores usage (%)       | 1,107                                                             |
| Working Set (MB)      | 199                                                               |
| Build Time (ms)       | 3,230                                                             |
| Start Time (ms)       | 376                                                               |
| Published Size (KB)   | 103,364                                                           |
| .NET Core SDK Version | 6.0.100-preview.2.21118.23                                        |
| ASP.NET Core Version  | 6.0.0-preview.2.21118.14+7a0e7fef3b4d32a42a014e6de2652034a37952f5 |
| .NET Runtime Version  | 6.0.0-dev                                                         |


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 30        |
| Cores usage (%)        | 363       |
| Working Set (MB)       | 38        |
| Start Time (ms)        | 1         |
| First Request (ms)     | 547       |
| Requests/sec           | 110,642   |
| Requests               | 1,670,616 |
| Mean latency (ms)      | 2.38      |
| Max latency (ms)       | 47.05     |
| Bad responses          | 0         |
| Socket errors          | 0         |
| Read throughput (MB/s) | 144.24    |
| Latency 50th (ms)      | 2.24      |
| Latency 75th (ms)      | 2.55      |
| Latency 90th (ms)      | 2.89      |
| Latency 99th (ms)      | 6.55      |

Reply:
Was there an anomaly in the JSON run? It didn't drop very much, but it would be nice to run it again just to make sure we didn't regress it.

### master


Requests/sec | 350,482
-- | --
Requests | 5,292,163
Mean latency (ms) | 1.95
Max latency (ms) | 202.26

### this PR


Requests/sec | 344,970
-- | --
Requests | 5,208,814
Mean latency (ms) | 1.96
Max latency (ms) | 311.77

Reply:
I'll run it again to be sure.

Reply:
Yeah it looks like it was just an anomaly. I ran it again against this PR and got these results (better than master):

| application           |                                                                  |
| --------------------- | ---------------------------------------------------------------- |
| CPU Usage (%)         | 78                                                               |
| Cores usage (%)       | 932                                                              |
| Working Set (MB)      | 151                                                              |
| Build Time (ms)       | 12,100                                                           |
| Start Time (ms)       | 298                                                              |
| Published Size (KB)   | 103,364                                                          |
| .NET Core SDK Version | 6.0.100-preview.2.21119.3                                        |
| ASP.NET Core Version  | 6.0.0-preview.2.21119.2+01982deeb026ed635ca96ab6a48e2028f53120ed |
| .NET Runtime Version  | 6.0.0-dev                                                        |


| load                   |           |
| ---------------------- | --------- |
| CPU Usage (%)          | 70        |
| Cores usage (%)        | 845       |
| Working Set (MB)       | 38        |
| Start Time (ms)        | 1         |
| First Request (ms)     | 126       |
| Requests/sec           | 356,321   |
| Requests               | 5,380,445 |
| Mean latency (ms)      | 1.63      |
| Max latency (ms)       | 170.36    |
| Bad responses          | 0         |
| Socket errors          | 0         |
| Read throughput (MB/s) | 51.65     |
| Latency 50th (ms)      | 0.68      |
| Latency 75th (ms)      | 0.79      |
| Latency 90th (ms)      | 0.94      |
| Latency 99th (ms)      | 35.90     |